File: store_client_credentials.py
#!/usr/bin/env python3
"""
Script para almacenar credenciales de clientes en Firestore
Esto permite que el admin panel muestre usuario y contraseña
"""
import firebase_admin
from firebase_admin import credentials, firestore, auth
import os

# Inicializar Firebase
try:
    app = firebase_admin.get_app()
except ValueError:
    cred = credentials.Certificate('credentials.json') if os.path.exists('credentials.json') else None
    if cred:
        firebase_admin.initialize_app(cred)
    else:
        # Si no hay credentials.json, usar credenciales por defecto
        firebase_admin.initialize_app()

db = firestore.client()

def store_client_credentials(client_id, email, password):
    """Almacenar credenciales de cliente en Firestore"""
    try:
        # Obtener documento del cliente
        doc = db.collection('clientes').document(client_id).get()
        
        if not doc.exists:
            print(f"❌ Cliente {client_id} no existe")
            return False
        
        # Actualizar con credenciales
        db.collection('clientes').document(client_id).update({
            'email_admin': email,
            'contraseña': password,  # NOTA: Idealmente usar bcrypt
            'password_plain': password,  # Guardado en texto plano para mostrar (¡SEGURIDAD!)
            'url_acceso': 'https://trafico-map-general-v2.web.app/client/'
        })
        
        print(f"✅ Credenciales almacenadas para {client_id}")
        print(f"   Email: {email}")
        print(f"   Contraseña: {password}")
        return True
        
    except Exception as e:
        print(f"❌ Error: {e}")
        return False

def get_all_clients_and_update():
    """Obtener todos los clientes y asegurar que tengan credenciales"""
    try:
        print("\n📋 Obteniendo todos los clientes...\n")
        
        clientes = db.collection('clientes').get()
        
        count = 0
        for doc in clientes:
            count += 1
            client_id = doc.id
            data = doc.data()
            
            print(f"🔍 Cliente #{count}: {data.get('nombre')} ({client_id})")
            
            # Verificar si ya tiene credenciales
            if 'email_admin' in data and 'contraseña' in data:
                print(f"   ✅ Ya tiene credenciales")
                print(f"   Email: {data['email_admin']}")
                print(f"   Contraseña: {data['contraseña']}")
            else:
                print(f"   ⚠️ Falta agregar credenciales")
                # El email del cliente
                email = data.get('email', f'admin@{client_id}.local')
                # Generar contraseña por defecto si no existe
                password = data.get('contraseña', f'Temporal{client_id}2024!')
                
                if store_client_credentials(client_id, email, password):
                    print(f"   ✅ Credenciales agregadas\n")
                else:
                    print(f"   ❌ Error agregando credenciales\n")
            
            print()
        
        if count == 0:
            print("❌ No se encontraron clientes en la colección")
        else:
            print(f"\n✅ Procesados {count} clientes")
        
    except Exception as e:
        print(f"❌ Error: {e}")
        import traceback
        traceback.print_exc()

if __name__ == "__main__":
    print("=" * 60)
    print("📋 ALMACENANDO CREDENCIALES DE CLIENTES EN FIRESTORE")
    print("=" * 60)
    
    get_all_clients_and_update()
    
    print("\n" + "=" * 60)
    print("✅ PROCESO COMPLETADO")
    print("=" * 60)

File: store-client-credentials.js
#!/usr/bin/env node
/**
 * Script para guardar credenciales de clientes en Firestore
 * Ejecución: node store-client-credentials.js
 */

import admin from 'firebase-admin';
import { fileURLToPath } from 'url';
import { dirname, join } from 'path';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

// Inicializar Firebase
try {
  admin.initializeApp({
    projectId: 'trafico-map-general-v2'
  });
} catch (e) {
  // Ya inicializado
}

const db = admin.firestore();

async function storeClientCredentials(clientId, email, password, clientNombre) {
  try {
    console.log(`📝 Guardando credenciales para ${clientId}...`);
    
    // Extraer nombre normalizado del cliente para la URL
    let urlNombre = clientId;
    
    // Si tenemos el nombre del cliente, usamos una versión simplificada
    if (clientNombre) {
      // Normalizar el nombre: convertir a minúsculas, reemplazar espacios por guiones
      urlNombre = clientNombre
        .toLowerCase()
        .trim()
        .replace(/\s+/g, '-')
        .replace(/á/g, 'a')
        .replace(/é/g, 'e')
        .replace(/í/g, 'i')
        .replace(/ó/g, 'o')
        .replace(/ú/g, 'u');
    }
    
    const url_acceso = `https://cliente-${urlNombre}.web.app/client/`;
    
    await db.collection('clientes').doc(clientId).update({
      email_admin: email,
      contraseña: password,
      password_plain: password,
      url_acceso: url_acceso,
      credenciales_actualizadas: new Date()
    });
    
    console.log(`✅ Credenciales guardadas`);
    console.log(`   Email: ${email}`);
    console.log(`   Contraseña: ${password}`);
    console.log(`   URL: ${url_acceso}\n`);
    return true;
  } catch (error) {
    console.error(`❌ Error: ${error.message}`);
    return false;
  }
}

async function getAllClientsAndUpdate() {
  try {
    console.log('\n📋 Obteniendo todos los clientes...\n');
    
    const snapshot = await db.collection('clientes').get();
    
    if (snapshot.empty) {
      console.log('❌ No se encontraron clientes');
      return;
    }
    
    console.log(`✅ Encontrados ${snapshot.size} clientes\n`);
    
    let updated = 0;
    for (const doc of snapshot.docs) {
      const clientId = doc.id;
      const data = doc.data();
      
      console.log(`🔍 Cliente: ${data.nombre} (${clientId})`);
      
      // Verificar si ya tiene credenciales
      if (data.email_admin && data.contraseña) {
        console.log(`   ✅ Ya tiene credenciales`);
        console.log(`   Email: ${data.email_admin}`);
        console.log(`   Contraseña: ${data.contraseña}`);
        console.log(`   URL: ${data.url_acceso}\n`);
      } else {
        const email = data.email || `admin@${clientId}.local`;
        const password = `Temporal${clientId}2024!`;
        
        if (await storeClientCredentials(clientId, email, password, data.nombre)) {
          updated++;
        }
      }
    }
    
    console.log(`\n✅ COMPLETADO - ${updated} clientes actualizados`);
    
  } catch (error) {
    console.error(`❌ Error: ${error.message}`);
    console.error(error);
  }
}

// Ejecutar
console.log('= '.repeat(40));
console.log('📋 ALMACENANDO CREDENCIALES DE CLIENTES');
console.log('= '.repeat(40));

getAllClientsAndUpdate().then(() => {
  console.log('\n= '.repeat(40));
  process.exit(0);
}).catch(error => {
  console.error(error);
  process.exit(1);
});
